### PR TITLE
Fake Join/Quit Messages for Vanish

### LINKED ISSUE #3880: Show join and quit messages when a player vanishes
### Feature request

**Feature description:**  
Two config options called something like message-on-vanish and message-on-unvanish which are broadcasted when vanish is turned on and off respectively

**How the feature is useful:**  
This can be used to make fake join and quit messages. A similar feature was requested earlier (https://github.com/EssentialsX/Essentials/issues/2713) but was closed because there wasn't a reliable way to get join messages. This should fix that issue by allowing the server owners to input the format themselves.

### PR DESCRIPTION
**Pull request by akdukaan**
Closes https://github.com/EssentialsX/Essentials/issues/3880

Adds the config option below:
```yml
# When a player enters/exits vanish, should a fake login/logout message be sent?
send-vanish-broadcast: false
```

The message is taken from messages.properties which defaults to the same as vanilla:
```properties
unvanishBroadcast=\u00a7e{0} joined the game
vanishBroadcast=\u00a7e{0} left the game
```

The messages can use `{0}` for `player.getName()` or `{1}` for `player.getDisplayName()`.

### CODE REVIEW

**Comment by Chew** [comment]
As stated in #3880 these shouldn't be strings. If you're wanting to vanish to simulate a fake join/quit, they should match what's already been configured earlier in the config. Them not matching ruins the entire point.

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandvanish.java [hunk 1]
```diff
@@ -40,6 +45,30 @@ void togglePlayer(final CommandSource sender, final User user, Boolean enabled)
...
         }
+
+        if (ess.getSettings().isFakeMessageOnVanish()) {
+            final Player player = user.getBase();
+            String msg;
+            if (enabled) {
+                if (ess.getSettings().isCustomQuitMessage()) {
+                    msg = ess.getSettings().getCustomQuitMessage();
+                } else {
+                    msg = "&e{USERNAME} left the game.";
```
Comment (Octol1ttle) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandvanish.java]:
```suggestion
                    msg = "&e{USERNAME} left the game";
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandvanish.java [hunk 2]
```diff
@@ -40,6 +45,30 @@ void togglePlayer(final CommandSource sender, final User user, Boolean enabled)
...
+                }
+            } else {
+                if (ess.getSettings().isCustomJoinMessage()) {
+                    msg = ess.getSettings().getCustomJoinMessage();
+                } else {
+                    msg = "&e{USERNAME} joined the game.";
```
Comment (Octol1ttle) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandvanish.java]:
```suggestion
                    msg = "&e{USERNAME} joined the game";
```

**Comment by Octol1ttle** [comment]
PSA: This implementation of fake join/quit messages **does not perfectly replicate Vanilla** due to it not using the new advanced chat stuff. If you hover on a player name in any of normal (not custom) join/quit messages, you will get a pop-up. If you hover on the fake messages, you won't get anything, meaning that anyone can easily spot whether someone vanished or not.
This isn't an issue with custom join/quit messages, as they also don't feature advanced chat stuff

Also a question: is this really something that needs to be added to Essentials? I feel like it would be better to just get a vanish plugin for advanced stuff like this

**Comment by akdukaan** [comment]
Yes, I feel that this really is something that needs to be added to Essentials. Essentials already provides basic vanish functionality, and fake quit messages are a basic aspect of vanish. Players can easily tell when a player enters vanish if there is no quit message. Plus, Essentials already has an option for joining in vanish which I feel is more advanced than fake quit messages.

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandvanish.java [hunk 3]
```diff
@@ -40,6 +45,30 @@ void togglePlayer(final CommandSource sender, final User user, Boolean enabled)
...
+                if (ess.getSettings().isCustomQuitMessage()) {
+                    msg = ess.getSettings().getCustomQuitMessage();
+                } else {
+                    msg = "&e{USERNAME} left the game";
+                }
+            } else {
+                if (ess.getSettings().isCustomJoinMessage()) {
+                    msg = ess.getSettings().getCustomJoinMessage();
+                } else {
+                    msg = "&e{USERNAME} joined the game";
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandvanish.java]: must be configurable, plugins may alter the join/leave message.

**Review by JRoy — CHANGES_REQUESTED** [review]

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandvanish.java (hunk 3 above)
Comment (akdukaan) [reply Essentials/src/main/java/com/earth2me/essentials/commands/Commandvanish.java]: That's how it was at https://github.com/EssentialsX/Essentials/pull/3881/commits/1d5b94f7fe4159d61e40968586d2f447a3cad119. Should I revert to the way it was at that point or is there a better way to make them configurable?

EDIT: Thanks Larry for answering me on the Discord. I'll start on the changes soon.

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandvanish.java [hunk 4]
```diff
@@ -40,6 +41,19 @@ void togglePlayer(final CommandSource sender, final User user, Boolean enabled)
...
         }
+
+        final Player player = user.getBase();
+        final boolean vanishBroadcast = ess.getSettings().vanishBroadcast();
+        if (vanishBroadcast) {
+            String broadcastMessage = "";
+            if (enabled) {
+                broadcastMessage = tl("vanishBroadcast", player.getName(), player.getDisplayName());
+            } else {
+                broadcastMessage = tl("unvanishBroadcast", player.getName(), player.getDisplayName());
```
Comment (pop4959) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandvanish.java:51]: Should use the actual join/quit message here, not hard-coded to tl, as other plugins will modify it

**Review by pop4959 — CHANGES_REQUESTED** [review]

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandvanish.java (hunk 4 above)
Comment (imDaniX) [reply Essentials/src/main/java/com/earth2me/essentials/commands/Commandvanish.java:51]: Is it possible tho? Plugins can modify messages only on firing of specific events, but firing join/quit event may lead to huge problems. The only way I see is to add the message into `VanishStatusChangeEvent` and hope that other plugins/developers with hook into it.

File: Essentials/src/main/java/com/earth2me/essentials/ISettings.java [hunk 5]
```diff
@@ -275,6 +275,8 @@
 
     boolean allowSilentJoinQuit();
 
+    boolean vanishBroadcast();
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/ISettings.java]:
```suggestion
    boolean isFakeVanishBroadcast();
```

File: Essentials/src/main/java/com/earth2me/essentials/Settings.java [hunk 6]
```diff
@@ -1233,6 +1233,11 @@ public boolean allowSilentJoinQuit() {
         return allowSilentJoin;
     }
 
+    @Override
+    public boolean vanishBroadcast() {
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/Settings.java]:
```suggestion
    public boolean isFakeVanishBroadcast() {
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandvanish.java [hunk 7]
```diff
@@ -40,6 +41,19 @@ void togglePlayer(final CommandSource sender, final User user, Boolean enabled)
         if (enabled) {
             user.sendMessage(tl("vanished"));
         }
+
+        final Player player = user.getBase();
+        final boolean vanishBroadcast = ess.getSettings().vanishBroadcast();
+        if (vanishBroadcast) {
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandvanish.java]:
```suggestion
        if (ess.getSettings().isFakeVanishBroadcast()) {
```

**Review by JRoy — CHANGES_REQUESTED** [review]
I've thought about it and think this can best be accepted by adding a field to `VanishStatusChangeEvent` that allows plugins to change the join/leave message.

The methods added should be as follows;
`isFakeJoinMessage()` - Is this vanish change event causing a fake join message
`isFakeLeaveMessage()` - Is this vanish change event causing a fake leave message
`getFakeJoinMessage()` - Gets the default fake join message from translation if `isFakeJoinMessage` returns true
`getFakeLeaveMessage()` - Gets the default fake leave message from translation if `isFakeLeaveMessage` returns true
`setFakeJoinMessage()` - Set's the fake join message
`setFakeLeaveMessage()` - Set's the fake leave message

Sorry for the constant flip flop in reviews on this PR, took me a while to think of something that would work best. If complete the spec as per above, I'll merge this PR

File: Essentials/src/main/java/net/ess3/api/events/VanishStatusChangeEvent.java [hunk 8]
```diff
@@ -24,4 +28,52 @@ public HandlerList getHandlers() {
...
+        }
+        return null;
+    }
+
+    /**
+     * @param message Sets the fake leave message.
+     */
+    public void setFakeLeaveMessage(String message) {
+        fakeLeaveMessage = message;
+    }
```
Comment (imDaniX) [inline Essentials/src/main/java/net/ess3/api/events/VanishStatusChangeEvent.java:78]: Can't it be merged after all?
`hasFakeMessage()` - is this event causing the message to show up.
`getFakeMessage()` - returns the message.
`setFakeMessage(String message)` - sets the message.
Cuz `StatusChangeEvent#getValue` already returns either it's "join" or "exit"

**Comment by pop4959** [comment]
This was discussed more internally and we decided not to merge this PR as there are still quite a few issues with it, which for the most part are not addressable at the moment:

- The fake join and quit messages will not match those set by other plugins, and would need to be manually updated. Even as an optional feature, users would expect this to work out of the box when enabled and not have to hack around translation messages to get this to work properly.
- There is no support for hover, etc, as mentioned by l1ttleO in a previous comment, making it obvious in most cases for an observant player to tell you are vanished. Obviously this is a limitation in Essentials itself currently and should be addressed by #2391.
- The changes suggested by JRoy above are a bit too idealistic, since it requires *all* other plugins to play nice with Essentials and hook in to provide their join / quit messages, and that's just not going to happen.
- Essentials vanish already as-is does not perfectly hide players, as it uses the Bukkit API instead of hiding players at a packet level. This is something which is out of scope and is best handled by a dedicated vanish plugin such as the classic [VanishNoPacket](https://dev.bukkit.org/projects/vanish).

That being said, thank you again for your contribution.

**Pull request closed without merging**